javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,890 @@
+    } else {
+      offset = 0
+      _String.lastCharset = encoding
+
+      val charBuffer = encoding.decode(ByteBuffer.wrap(data, start, length))
+      value = charBuffer.array
+      count = charBuffer.length
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: _String) =
+    this(data, start, length, Charset.forName(encoding))
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) =
+    this(data, start, length, Charset.defaultCharset())
+
+  def this(data: Array[scala.Byte], high: Int) =
+    this(data, high, 0, data.length)
+
+  def this(data: Array[scala.Byte], encoding: _String) =
+    this(data, 0, data.length, encoding)
+
+  def this(data: Array[scala.Byte], encoding: Charset) =
+    this(data, 0, data.length, encoding)
+
+  def this(data: Array[scala.Byte]) =
+    this(data, 0, data.length)
+
+  def this(data: Array[Char], start: Int, length: Int) = {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      value = new Array[Char](length)
+      count = length
+      System.arraycopy(data, start, value, 0, count)
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(data: Array[Char]) =
+    this(data, 0, data.length)
+
+  def this(start: Int, length: Int, data: Array[Char]) = {
+    this()
+    value = data
+    offset = start
+    count = length
+  }
+
+  def this(string: _String) = {
+    this()
+    value = string.value
+    offset = string.offset
+    count = string.length()
+  }
+
+  def this(sb: StringBuffer) {
+    this()
+    offset = 0
+    synchronized {
+      value = sb.getValue
+      count = sb.length
+    }
+  }
+
+  def this(codePoints: Array[Int], offset: Int, count: Int) {
+    this()
+    ???
+//    if (offset < 0 || count < 0 || offset > codePoints.length - count) {
+//      throw new IndexOutOfBoundsException()
+//    } else {
+//      this.offset = 0
+//      this.value = new Array[Char](count * 2)
+//      val end = offset + count
+//      var c = 0
+//
+//      var i = offset
+//      while (i < end) {
+//        c +=
+//          Character.toChars(codePoints(i), this.value, c)
+//      }
+//
+//      this.count = c
+//    }
+  }
+
+  def this(sb: java.lang.StringBuilder) {
+    this()
+    offset = 0
+    count = sb.length
+    value = new Array[Char](count)
+    sb.getChars(0, count, value, 0)
+  }
+
+  def charAt(index: Int): Char = {
+    if (0 <= index && index < count) {
+      value(offset + index)
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  private def compareValue(ch: Char): Char = {
+    if (ch < 128) {
+      if ('A' <= ch && ch <= 'Z') (ch + ('a' - 'A')).toChar
+      else ch
+    } else {
+      Character.toLowerCase(Character.toUpperCase(ch))
+    }
+  }
+
+  private def toLowerCase(ch: Char): Char = {
+    if (ch < 128) {
+      if ('A' <= ch && ch <= 'Z') (ch + ('a' - 'A')).toChar
+      else ch
```

Comment:
Looks like these micro-optimizations are best left to `Character.toLowerCase()` itself.